src/charm.py

```diff
@@ -399,8 +405,15 @@ def _on_start(self, _):
 
     def _on_config_changed(self, _):
         """Event handler for ConfigChangedEvent."""
+        self._save_ssh_key()
         self._common_exit_hook()
 
+    def _save_ssh_key(self):
+        """Save SSH key from config to a file."""
+        ssh_key = self.config.get("git_ssh_key")
+        if ssh_key:
+            Path(self._ssh_key_file_name).write_text(ssh_key)
```

Comment:
**suggestion:** write the file anyway. If we had one before, but no longer do, we should probably empty it out (idempotent).

```suggestion
        ssh_key = self.config.get("git_ssh_key", "")
        Path(self._ssh_key_file_name).write_text(ssh_key)
```

Also, iiuc, we need `pebble.push` here, with an absolute path.
Setting up a webserver for unittesting is a pain, but perhaps you could manually test it with a dummy repo?

Reply:
+1

Reply:
Fortunately (maybe) there's already a [threaded mocked webserver in one of our unit tests](https://github.com/canonical/loki-k8s-operator/blob/main/tests/unit/test_loki_server.py) as an example